rosbag2_storage_default_plugins/src/rosbag2_storage_default_plugins/sqlite/sqlite_storage.cpp

```diff
@@ -48,43 +48,29 @@ namespace rosbag2_storage_plugins
 void SqliteStorage::open(
   const std::string & uri, rosbag2_storage::storage_interfaces::IOFlag io_flag)
 {
-  auto metadata = is_read_only(io_flag) ?
-    load_metadata(uri) :
-    std::unique_ptr<rosbag2_storage::BagMetadata>();
-
-  if (metadata) {
-    if (metadata->relative_file_paths.empty()) {
-      throw std::runtime_error(
-              "Failed to read from bag '" + uri + "': Missing database file path in metadata");
-    }
-
-    database_name_ = metadata->relative_file_paths[0];
-  } else {
-    if (is_read_only(io_flag)) {
-      throw std::runtime_error("Failed to read from bag '" + uri + "': No metadata found.");
-    }
-
-    database_name_ = rosbag2_storage::FilesystemHelper::get_folder_name(uri) + ".db3";
-  }
+  relative_path_ = uri + ".db3";
```

Comment:
Looking at this again, I think this is not the correct approach. A call to `open()` should not compose a URI. 

In the existing implementation, we could do that because of loading the metadata at the same time. I think the write approach here is that we introduce a new function on the storage interfaces which might give you a plugin specific file extension or something which then allows you to create the individual uri in the Writer API, i.e. `format_storage_uri()`.

In both cases, read and write, the passed in `uri` should point to a valid sqlite database file name. In the case of  writing a new file, this database will be created as given in the uri.

Reply:
Moving the composition of the `uri` into a separate function will require changes to `StorageFactory` and `StorageFactoryImpl` since `Writer` does not know what the extension should be until after the storage interface has been opened.

Reply:
An option I have considered is to have `SqliteStorage::open` accept `uri` with and without the extension. In this case, `Writer` can continue to open the plugin via `StorageFactory` without knowing the extension and `Reader` can expect the storage plugin to open the bagfile with the relative file path obtained from the metadata.

@Karsten1987  would this suggestion work for you?

Reply:
I think that would work. I think we can assume that every time we pass in `IOFlag::READ_WRITE` a new file is created - opposed to `IOFlag::APPEND`. see https://github.com/ros2/rosbag2/blob/master/rosbag2_storage/include/rosbag2_storage/storage_interfaces/base_io_interface.hpp#L30-L31

In the case of `IOFlag::READ_WRITE` the `.db3` suffix will be added.

Reply:
Ok, I'll make the change to apply `.db3` to the `uri` when `READ_WRITE` is passed.

Reply:
I've applied this change.

`SqliteStorage::open(uri, READ_WRITE)` will append the extension, since `Writer` calls this indirectly through `StorageFactory`, before it knows what extension the `db` should be.

`SqliteStorage::open(uri, READ_ONLY)` expects uri to have the extension since `Reader` calls this by passing a `relative_file_path` read from metadata.

`APPEND` behaves similarly to `READ_ONLY`.